bundles/org.eclipse.e4.ui.workbench.compatibility.migration/src/org/eclipse/e4/ui/workbench/compatibiliy/migration/Converter.java

```diff
@@ -0,0 +1,933 @@
+import static org.eclipse.ui.internal.IWorkbenchConstants.TAG_WINDOW;
+import static org.eclipse.ui.internal.IWorkbenchConstants.TAG_WORKBOOK;
+import static org.eclipse.ui.internal.IWorkbenchConstants.TAG_WORKING_SET;
+import static org.eclipse.ui.internal.IWorkbenchConstants.TAG_WORKING_SETS;
+import static org.eclipse.ui.internal.IWorkbenchConstants.TAG_X;
+import static org.eclipse.ui.internal.IWorkbenchConstants.TAG_Y;
+
+import java.io.BufferedReader;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.StringWriter;
+import java.lang.reflect.Field;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.e4.ui.model.application.MAddon;
+import org.eclipse.e4.ui.model.application.MApplication;
+import org.eclipse.e4.ui.model.application.MApplicationFactory;
+import org.eclipse.e4.ui.model.application.ui.MUIElement;
+import org.eclipse.e4.ui.model.application.ui.advanced.MAdvancedFactory;
+import org.eclipse.e4.ui.model.application.ui.advanced.MArea;
+import org.eclipse.e4.ui.model.application.ui.advanced.MPerspective;
+import org.eclipse.e4.ui.model.application.ui.advanced.MPerspectiveStack;
+import org.eclipse.e4.ui.model.application.ui.advanced.MPlaceholder;
+import org.eclipse.e4.ui.model.application.ui.basic.MBasicFactory;
+import org.eclipse.e4.ui.model.application.ui.basic.MPart;
+import org.eclipse.e4.ui.model.application.ui.basic.MPartSashContainer;
+import org.eclipse.e4.ui.model.application.ui.basic.MPartSashContainerElement;
+import org.eclipse.e4.ui.model.application.ui.basic.MPartStack;
+import org.eclipse.e4.ui.model.application.ui.basic.MStackElement;
+import org.eclipse.e4.ui.model.application.ui.basic.MTrimmedWindow;
+import org.eclipse.e4.ui.model.application.ui.basic.MWindow;
+import org.eclipse.e4.ui.workbench.IPresentationEngine;
+import org.eclipse.e4.ui.workbench.modeling.EPartService;
+import org.eclipse.e4.ui.workbench.persistence.common.CommonUtil;
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.resource.Resource;
+import org.eclipse.emf.ecore.resource.ResourceSet;
+import org.eclipse.emf.ecore.resource.impl.ResourceSetImpl;
+import org.eclipse.swt.graphics.Rectangle;
+import org.eclipse.ui.IMemento;
+import org.eclipse.ui.WorkbenchException;
+import org.eclipse.ui.XMLMemento;
+import org.eclipse.ui.internal.e4.compatibility.ModeledPageLayoutUtils;
+
+/**
+ * The Implementation of the Converter interface.
+ *
+ * @since 3.3
+ *
+ */
+@SuppressWarnings("restriction")
+public class Converter {

```

Comment:
Name is a bit weak, maybe something more explicit `E4MigrationTool` ?

Reply:
Or maybe this class is not intended to be API and it could be package visible? In such case, we can care much less about documentation and API names.